### ISSUE TITLE
dockerize backend

### ISSUE THREAD

**Issue by themightychris** [CodeForPhilly/prevention-point#223]
To deploy to the civic cloud, the backend server will need to be dockerized as #212 covers for the frontend

This may just be a matter of copying the [Getting Started / Back-end steps from the README](https://github.com/CodeForPhilly/prevention-point#back-end) into a Dockerfile with an appropriate base image, and integrating it with the DB and frontend integrated in #218 

The project needs to be gotten to the point where running `docker-compose up` inside a fresh clone gets a fully-configured instance of the app exposed on an HTTP port

**Comment by ThrowsException** [comment]
Just to clarify what exactly needs to be exposed? Im obviously not familiar with how the production deployment of this is intending to be setup but is the goal to run `docker-compose up -d` in production and have a server running the docker compose or is the docker-compose simply for development and the deployment will be the individual docker containers running in seperate environments.

**Comment by themightychris** [comment]
### TL;DR:

If you get it working however is convenient for the internal development team through `docker-compose up`, the CfP ops team can handle making any adaptations needed

Ideally, env vars can drive hostnames (for TCP links) or base URLs (ideally for all HTTP links so scheme/host/path can vary by deployment) for components to talk to each other. If you need them known at build time, make them build args and ops team will have to figure something out

Where it doesn't make development inconvenient, we'd prefer to see a single container exposing backend+frontend over one HTTP port, but it's not worth making dev workflows more complex over if it goes against the grain too much in your stack.

---

### More background

The `docker-compose` will serve two purposes:

1. ensure development is accessible for future maintainers (both inside and outside the project team)
2. give the CfP central ops team a definitive and maintained reference point for how to put the application together in a container environment

What #2 looks like is an evolving target that's in pretty early stages right now. At the moment, we're exploring how we might use ansible to power [a collection of k8s templates](https://github.com/CodeForPhilly/civic-cluster/tree/master/admin/templates/project) that bundle up best practices for common project stacks, so we can keep [project deployments](https://github.com/CodeForPhilly/civic-cluster/tree/master/projects) as slim in complexity as possible.

On the ops end we're definitely still many iterations away from what a stable system will look like, so we're trying to isolate project dev teams from worrying about that by making `docker-compose` the mediator. We settled on that because it's both a best practice for within the team to make development most accessible, and then in fulfilling that role it's guaranteed to stay an up-to-date reference for the external ops team to create/maintain/refactor deployment recipes as our infra evolves.

For the project team, we'd ideally like to see a `docker-compose`-driven workflow become your primary path for developers within the team so that we can guarantee it's maintained, and increase the odds that years later when a new project team needs to pick things up to make enhancements they don't get stuck navigating how current versions of all the dependent tools can't run the app anymore

**Comment by themightychris** [comment]
Some further notes from [a Slack thread](https://codeforphilly.slack.com/archives/CGHQ130MQ/p1581470864155700?thread_ts=1581467654.154700&cid=CGHQ130MQ) re: database

> ops team will handle setting up a backup process, the docker-compose should just include its own postgres image. We might look to the project team for guidance on the appropriate backup interval (e.g. hourly?) Long term we’ll probably want to replace that on our end with a managed database of some sort

**Comment by ThrowsException** [comment]
Couple thoughts I have on this. 
First I wouldn't consider docker-compose to to be a "dev workflow". trying to build containers that facilitate live development ends up falling flat because while developing you need far more tools such as proxies, linters, build-tools etc that the containers become bloated and fragile and not reflective or a production application. Running `docker-compose up -d` should however still be a good indicator of a working project. 

Secondly it makes communication way more complex. This is where things like CORs become issues. for example 

client-container:8080 
backend-container:8000. 

Visiting the exposed client applications loads the client app but has no way to communicate with the backend without configuring COR's because of the port numbers or routing through a third container like nginx using docker networking like 

nginx-container:8000/ -> frontend-contatiner:8080 
nginx-container:8000/api -> backend-container:8000

My suggestion here to see a running application would be to do something like ive outlined with the nginx container so someone wanting to use or see the project could simply `docker-compose up -d` and the entire project is running. For front end devs provides scripts to only start the database and backend service and rely on `npm start` for local dev. For backend work as normal with the provided instrcutions. `docker-compose up` should just be a final check to make sure the container still builds and runs. 

I can try to get this all into one PR if that makes sense and vet my idea here.

**Comment by themightychris** [comment]
@ThrowsException thanks for all your thinking on this and taking the time to write so much up! It all sounds spot-on to me, I'd love to see what your thinking looks like in practice through a PR

One variation I'm thinking about, but won't have a strong opinion on until we can consider what the difference actually looks like in practice, is a model where the `docker-compose` configuration and associated documentation is kept as simple as possible, focused on providing a minimal complete environment from working tree. I don't know what your approach will look like, but I'm thinking about other efforts I've seen to maintain multiple targets within docker-compose become very complex.

So a single run target, but then a second tier of "expert mode" paths for frontend or backend developers where in a second step you can run a more debuggable frontend or backend and put it in front of the container version through env/query params. This keeps docker-compose in a form that new devs and external ops will get the most useful understanding out of reading through.

This approach is a bit more time and resource inefficient. You'll be running possibly an extra container, but you've already got the container engine running. You might have to wait for container build that's redundant with the dev-mode build your about to do, but Docker's caching will help out. Then on the plus side, for whatever someone is trying to work on they'll always have a "reference instance" to compare against as needed.

**Comment by mariekers** [comment]
How's this going @ThrowsException @themightychris - need anything?

**Comment by themightychris** [comment]
@ThrowsException I suggested some improvements: https://github.com/ThrowsException/prevention-point/pull/1

If those get merged in first than this looks good to merge to me :+1:

**Comment by ThrowsException** [comment]
@themightychris I thought I had addressed the last comment with the script names. Let me know if there are other issues.

**Comment by themightychris** [comment]
That's all I found, I've started trying to kick over to ops for deployment, can you get all the improvements merged?

**Comment by ThrowsException** [comment]
@themightychris branch is up to date.

**Comment by ThrowsException** [comment]
Is this and #212 good to close? I feel like we've sufficiently addressed these for now. @themightychris
